File: backend/app/schemas/algorithm.py
from pydantic import BaseModel
from typing import Literal, Optional, List

Category = Literal["sorting", "searching", "data-structures", "graphs"]
Difficulty = Literal["easy", "medium", "hard"]

class ComplexityInfo(BaseModel):
    best: str
    average: str
    worst: str

class AlgorithmMeta(BaseModel):
    id: str
    name: str
    category: Category
    difficulty: Difficulty
    description: str
    timeComplexity: ComplexityInfo
    spaceComplexity: str
    pseudocode: Optional[List[str]] = None
    tags: Optional[List[str]] = None

File: frontend/src/types/algorithm.ts
export type Category = "sorting" | "searching" | "data-structures" | "graphs";

export type Difficulty = "easy" | "medium" | "hard";

export type ComplexityInfo = {
  best: string;
  average: string;
  worst: string;
};

export type AlgorithmMeta = {
  id: string;           // e.g. "bubble-sort"
  name: string;         // e.g. "Bubble Sort"
  category: Category;
  difficulty: Difficulty;
  description: string;
  timeComplexity: ComplexityInfo;
  spaceComplexity: string;
  pseudocode?: string[];
  tags?: string[];
};
